public/app/features/dashboard/components/ShareModal/SharePublicDashboard.tsx

```diff
@@ -1,60 +1,85 @@
 import React, { useState, useEffect } from 'react';
 
-import { Button, Field, Switch } from '@grafana/ui';
+import { Button, Field, Switch, Alert } from '@grafana/ui';
 import { notifyApp } from 'app/core/actions';
 import { createErrorNotification, createSuccessNotification } from 'app/core/copy/appNotification';
+import { contextSrv } from 'app/core/services/context_srv';
+import { DashboardModel } from 'app/features/dashboard/state/DashboardModel';
 import { dispatch } from 'app/store/store';
 
 import {
   dashboardCanBePublic,
   getPublicDashboardConfig,
   savePublicDashboardConfig,
+  getDashboard,
   PublicDashboardConfig,
+  DashboardResponse,
 } from './SharePublicDashboardUtils';
 import { ShareModalTabProps } from './types';
 
 interface Props extends ShareModalTabProps {}
 
-// 1. write test for dashboardCanBePublic
-// 2. figure out how to disable the switch
-
 export const SharePublicDashboard = (props: Props) => {
-  const [publicDashboardConfig, setPublicDashboardConfig] = useState<PublicDashboardConfig>({ isPublic: false });
   const dashboardUid = props.dashboard.uid;
+  const orgId = contextSrv.user.orgId;
+  const [publicDashboardConfig, setPublicDashboardConfig] = useState<PublicDashboardConfig>({
+    isPublic: false,
+    publicDashboard: { uid: '', dashboardUid, orgId },
+  });
+
+  const [dashboard, setDashboard] = useState<DashboardModel>({} as DashboardModel);
 
   useEffect(() => {
+    // dashboard model may be stale, so load it ourselves
+    getDashboard(dashboardUid)
+      .then((dashboardResponse: DashboardResponse) => {
+        setDashboard(new DashboardModel(dashboardResponse.dashboard, dashboardResponse.meta));
+      })
+      .catch((err) => {
+        dispatch(notifyApp(createErrorNotification('Failed to retrieve dashboard', err)));
+      });
```

Comment:
why are we loading the dashboard here? We have version column in dashboard for optimistic concurrency, so no need to do this (something dashboard_public_config should have as well to prevent user overwrites). 

But loading the public dashboard config should be enough here, no need to load the dashboard as well.

Reply:
When saving settings, the dashboard was not being reloaded into state properly. E.g. if there are not template variables, and you add them and save, the dashboard prop that is passed in dashnav -> sharemodal -> public dashboard panel never has that property updated. We tested this extensively yesterday and couldn't find a way to resolve it. This action was taken to keep us moving forward.

Reply:
@jalevin ok, sounds like a bug, but using props.dashboard.getVariables() should solve that.  What you are doing here causes the opposite problem, if the user has added variables and not saved, this check will still pass

Reply:
and the reason for this bug is that dashbord.templating.list is just the initial variable state, then it's handled in redux ( I know super messy), but props.dashboard.getVariables() should get the latest variable list from redux

Reply:
Thinking through this further, I think this actually poses a problem. If someone makes a change to template variables and doesn't persist the change, but goes and makes the dashboard public, we'll need to handle that validation on the backend and return an error. Is that the way we should handle this?